proxy/pkg/util/util.go

```diff
@@ -506,19 +511,257 @@ func getUserClusterList(projectList []string) []string {
+				if !ok {
+					namespaceCounts[namespace] = 1
+				} else {
+					namespaceCounts[namespace] = count + 1
+				}
+			}
+		}
+	}
+
+	klog.Infof("allAccessCount  : %v", allAccessCount)
+	klog.Infof("Namespaces Count  : %v", namespaceCounts)
+
+	commonNamespaces := []string{}
+
+	if allAccessCount == len(reqClustersMetricsAccess) {
+		commonNamespaces = append(commonNamespaces, "*")
+	} else {
+		for ns, count := range namespaceCounts {
+			if (count + allAccessCount) == len(reqClustersMetricsAccess) {
+				commonNamespaces = append(commonNamespaces, ns)
+			}
+		}
+	}
+
+	klog.Infof("common namespaces across clusters: %v   : %v", clusters, commonNamespaces)
+	return commonNamespaces
+}
+
+// read clusters in the user's query
+func getClustersInQuery(queryValues url.Values, key string, userMetricsAccess map[string][]string) []string {
+
+	// parse the promql query and return the queries  set of cluster
+	// so that better namespace filtering can be done
+
+	// for query  --  namespace:container_memory_usage_bytes:sum{cluster=~"devcluster1"}
+	// result should be -- "devcluster1"
+
+	// stubbing to return "empty" i.e all clusters
+
+	query := queryValues.Get(key)
+	reg := regexp.MustCompile(`([{|,][ ]*)cluster(=|!=|=~|!~)([ ]*)"([^"]+)"`)
+	matches := reg.FindStringSubmatch(query)
+	if len(matches) == 0 {
+		return []string{}
+	}
+	operator := matches[2]
+	expr := matches[4]
+	clusters := []string{}
+	for cluster := range userMetricsAccess {
+		clusters = append(clusters, cluster)
+	}
+	queryClusters := []string{}
+	switch operator {
+	case "=":
+		return []string{expr}
+	case "!=":
+		for _, cluster := range clusters {
+			if cluster != expr {
+				queryClusters = append(queryClusters, cluster)
+			}
+		}
+		return queryClusters
+	case "=~":
+		reg = regexp.MustCompile(expr)
+		for _, cluster := range clusters {
+			if reg.MatchString(cluster) {
+				queryClusters = append(queryClusters, cluster)
+			}
+		}
+		return queryClusters
+	case "!~":
+		reg = regexp.MustCompile(expr)
+		for _, cluster := range clusters {
+			if !reg.MatchString(cluster) {
+				queryClusters = append(queryClusters, cluster)
+			}
+		}
+		return queryClusters
+	}
+	return []string{}
+}
+
+func rewriteQuery(queryValues url.Values, userMetricsAccess map[string][]string, key string) url.Values {
+
 	originalQuery := queryValues.Get(key)
+
+	klog.Infof("REQRITE QUERY: key is: %v ,  originalQuery is: \n %v", key, originalQuery)
 	if len(originalQuery) == 0 {
 		return queryValues
 	}
 
+	//get clusters list for injecting cluster label
+	clusterList := make([]string, 0, len(userMetricsAccess))
+	for clusterName := range userMetricsAccess {
+		clusterList = append(clusterList, clusterName)
+	}
+
 	modifiedQuery, err := rewrite.InjectLabels(originalQuery, "cluster", clusterList)
 	if err != nil {
 		return queryValues
 	}
 
+	klog.Infof("REQRITE QUERY Modified Query after injecting clusters: \n %v", modifiedQuery)
+
+	modifiedQuery2 := modifiedQuery
```

Comment:
Refactored code to make it more readable. The metric  `acm_managed_cluster_labels` metric is treated special, as it requires the label `name` for cluster list instead of `cluster`,  and does not require injecting namespaces.